spring-boot-actuator/src/main/java/org/springframework/boot/actuate/health/DefaultHealthIndicatorRegistry.java

```diff
@@ -0,0 +1,149 @@
+ * limitations under the License.
+ */
+
+package org.springframework.boot.actuate.health;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.util.Assert;
+import org.springframework.util.ClassUtils;
+
+/**
+ * Default implementation of {@link HealthIndicatorRegistry}.
+ *
+ * @author Vedran Pavic
+ */
+public class DefaultHealthIndicatorRegistry implements HealthIndicatorRegistry {
+
+	private static final Log logger = LogFactory.getLog(DefaultHealthIndicatorRegistry.class);
+
+	private final Map<String, HealthIndicator> healthIndicators =
+			new HashMap<String, HealthIndicator>();
+
+	private DefaultHealthIndicatorRegistryProperties properties;
+
+	private ExecutorService executor;
+
+	/**
+	 * Create a {@link DefaultHealthIndicatorRegistry} instance.
+	 * @param properties the configuration properties
+	 */
+	public DefaultHealthIndicatorRegistry(DefaultHealthIndicatorRegistryProperties properties) {
+		Assert.notNull(properties, "Properties must not be null");
+		this.properties = properties;
+		if (this.properties.isRunInParallel()) {
+			this.executor = Executors.newFixedThreadPool(
+					this.properties.getThreadCount(), new WorkerThreadFactory());
+		}
+	}
+
+	/**
+	 * Create a {@link DefaultHealthIndicatorRegistry} instance.
+	 */
+	public DefaultHealthIndicatorRegistry() {
+		this(new DefaultHealthIndicatorRegistryProperties());
+	}
+
+	@Override
+	public void register(String name, HealthIndicator healthIndicator) {
+		synchronized (this.healthIndicators) {
+			this.healthIndicators.put(name, healthIndicator);
+		}
+	}
+
+	@Override
+	public void unregister(String name) {
+		synchronized (this.healthIndicators) {
+			this.healthIndicators.remove(name);
+		}
+	}
+
+	@Override
+	public Set<String> getRegisteredNames() {
+		return Collections.unmodifiableSet(new HashSet<String>(this.healthIndicators.keySet()));
+	}
+
+	@Override
+	public Health runHealthIndicator(String name) {
+		HealthIndicator healthIndicator = this.healthIndicators.get(name);
+		Assert.notNull(healthIndicator, "HealthIndicator " + name + " does not exist");
+		return healthIndicator.health();
+	}
+
+	@Override
+	public Map<String, Health> runHealthIndicators() {
+		Map<String, Health> healths = new HashMap<String, Health>();
+		if (this.properties.isRunInParallel()) {
+			Assert.notNull(this.executor, "Executor must not be null");
+			Map<String, Future<Health>> futures = new HashMap<String, Future<Health>>();
+			for (final Map.Entry<String, HealthIndicator> entry : this.healthIndicators.entrySet()) {
```

Comment:
This isn't thread-safe. To be safe, it shouldn't just be wrapped in a synchronised block. We should avoid calling a health indicator while holding a lock as it's code that we don't control and could result in deadlock if someone does something a bit daft in their indicator and calls back into the registry.